graph/iterator/resolver.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018 The Cayley Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"fmt"
+	"github.com/cayleygraph/cayley/graph"
+	"github.com/cayleygraph/cayley/quad"
+)
+
+var _ graph.Iterator = &Resolver{}
+
+// A Resolver iterator consists of it's nodes, an index (where it is in the,
+// process of iterating) and a store to resolve values from.
+type Resolver struct {
+	qs     graph.QuadStore
+	uid    uint64
+	tags   graph.Tagger
+	nodes  []quad.Value
+	index  int
+	err    error
+	result graph.Value
+}
+
+// Creates a new Resolver iterator.
+func NewResolver(qs graph.QuadStore, nodes ...quad.Value) *Resolver {
+	it := &Resolver{
+		uid:   NextUID(),
+		qs:    qs,
+		nodes: make([]quad.Value, 0, 20),
+	}
+	// Enforce uniqueness
+	unique := make(map[quad.Value]bool, 20)
```

Comment:
I would actually expect it to iterate over duplicates. It might resolve duplicates once, but in general, it should preserve an order and the number of values.